### PR TITLE
MODINREACH-100 - Borrower Renew Message for Item Hold (Owning Site)

### PR DESCRIPTION
**Pull request by s1mpleGit**
[MODINREACH-100](https://issues.folio.org/browse/MODINREACH-100) - D2IR Local Endpoint (Circulation): Borrower Renew Message for Item Hold (Owning Site)

## Purpose
Endpoint receives message from central server to owning site indicating that an item that was checked out to and received by the borrowing site (shipped) has been renewed by the borrowing site patron.

## Approach
Endpoints implemented
Tests added

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

### CODE REVIEW

File: src/main/java/org/folio/innreach/controller/InnReachTransactionController.java [hunk 1]
```diff
@@ -101,4 +102,13 @@ public InnReachResponseDTO handleMethodArgumentNotValidException(MethodArgumentN
...
   }
+
+  @Override
+  @PostMapping("/inn-reach/d2ir/circ/recall/{trackingId}/{centralCode}")
+  public ResponseEntity<InnReachResponseDTO> recallItem(@PathVariable String trackingId,
+                                                        @PathVariable String centralCode,
+                                                        InnReachRecallItemDTO recallItem) {
+    var response = transactionService.recallItem(trackingId, centralCode, recallItem);
+    return ResponseEntity.ok(response);
+  }
```
Comment (oliinyko) [inline src/main/java/org/folio/innreach/controller/InnReachTransactionController.java]: D2IR endpoints should be handled by `InnReachCirculationController`

File: src/main/java/org/folio/innreach/domain/service/InnReachTransactionService.java [hunk 2]
```diff
@@ -17,4 +18,6 @@
   InnReachTransactionDTO getInnReachTransaction(UUID transactionId);
 
   InnReachTransactionsDTO getAllTransactions(Integer offset, Integer limit, InnReachTransactionFilterParametersDTO parameters);
+
+  InnReachResponseDTO recallItem(String trackingId, String centralCode, InnReachRecallItemDTO recallItem);
```
Comment (oliinyko) [inline src/main/java/org/folio/innreach/domain/service/InnReachTransactionService.java:22]: should be moved to `CirculationService`

File: src/main/java/org/folio/innreach/controller/InnReachTransactionController.java (hunk 1 above)
Comment (oliinyko) [reply src/main/java/org/folio/innreach/controller/InnReachTransactionController.java]: By the way, it's not clear to me what's the purpose of adding this method? It seems to be out of scope for the current story

File: src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java [hunk 3]
```diff
@@ -262,6 +267,28 @@ public InnReachResponseDTO returnUncirculated(String trackingId, String centralC
...
+    var calculatedDueDate = transaction.getHold().getNeedBefore();
+    var requestedDueDate = borrowerRenew.getDueDateTime();
+
+    if (calculatedDueDate >= requestedDueDate) {
+      transaction.setState(BORROWER_RENEW);
+      transactionRepository.save(transaction);
+    } else {
+      try {
+        var recallItem = new InnReachRecallItemDTO().dueDateTime(calculatedDueDate);
+        transactionService.recallItem(trackingId, centralCode, recallItem);
```
Comment (oliinyko) [inline src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java]: > Issue a D2IR recall request to the central server using the current due date of the loan

There is no need to call any local endpoint. The requirement means that you need to call an endpoint of central server (i.e. by using `InnReachExternalService`)

File: src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java [hunk 4]
```diff
@@ -262,6 +267,28 @@ public InnReachResponseDTO returnUncirculated(String trackingId, String centralC
...
+      } catch (IllegalArgumentException e) {
+        var cancelRequest = new CancelRequestDTO().reason("Transaction cancelled at owning side");
+        cancelPatronHold(trackingId, centralCode, cancelRequest);
+      }
```
Comment (oliinyko) [inline src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java]: Is there a requirement to cancel transaction? I guess it's better to clarify with PO. 
Also, pay attention to the type of hold. The method `cancelPatronHold` is designed specifically for `Patron Hold` transactions while this one is `Item Hold`

File: src/main/java/org/folio/innreach/controller/InnReachTransactionController.java (hunk 1 above)
Comment (s1mpleGit) [reply src/main/java/org/folio/innreach/controller/InnReachTransactionController.java]: removed

File: src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java (hunk 4 above)
Comment (s1mpleGit) [reply src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java]: removed

File: src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java [hunk 5]
```diff
@@ -262,6 +267,26 @@ public InnReachResponseDTO returnUncirculated(String trackingId, String centralC
...
+  @Override
+  public InnReachResponseDTO borrowerRenew(String trackingId, String centralCode, BorrowerRenewDTO borrowerRenew) {
+    var transaction = getTransaction(trackingId, centralCode);
+    var transactionState = transaction.getState();
+    var calculatedDueDate = transaction.getHold().getNeedBefore();
+    var requestedDueDate = borrowerRenew.getDueDateTime();
+
+    if (calculatedDueDate >= requestedDueDate) {
+      transaction.setState(BORROWER_RENEW);
+      transactionRepository.save(transaction);
```
Comment (oliinyko) [inline src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java]: save operation is redundant here, the whole service is `@Transactional`

File: src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java [hunk 6]
```diff
@@ -262,6 +267,26 @@ public InnReachResponseDTO returnUncirculated(String trackingId, String centralC
...
+    } else if (transactionState == ITEM_SHIPPED) {
+      transaction.setState(RECALL);
+      transactionRepository.save(transaction);
```
Comment (oliinyko) [inline src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java]: same as above

File: src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java (hunk 5 above)
Comment (s1mpleGit) [reply src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java]: fixed

File: src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java (hunk 6 above)
Comment (s1mpleGit) [reply src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java]: fixed

File: src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java [hunk 7]
```diff
@@ -262,6 +267,26 @@ public InnReachResponseDTO returnUncirculated(String trackingId, String centralC
...
+      } else {
+      throw new IllegalArgumentException(UNEXPECTED_TRANSACTION_STATE + transactionState.name());
+    }
+
+    return success();
+  }
```
Comment (oliinyko) [inline src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java]: If I understand the requirements correctly we need to make a call to FOLIO circulation API to

> Attempt to renew the loan associated with the Item Hold transaction

File: src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java [hunk 8]
```diff
@@ -262,6 +267,26 @@ public InnReachResponseDTO returnUncirculated(String trackingId, String centralC
     }
   }
 
+  @Override
+  public InnReachResponseDTO borrowerRenew(String trackingId, String centralCode, BorrowerRenewDTO borrowerRenew) {
+    var transaction = getTransaction(trackingId, centralCode);
+    var transactionState = transaction.getState();
+    var calculatedDueDate = transaction.getHold().getNeedBefore();
+    var requestedDueDate = borrowerRenew.getDueDateTime();
```
Comment (oliinyko) [inline src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java]: > If **calculated due date** is < **requested due date**, or the renewal request fails and the **existing due date** is < the requested due date from the borrower

From this requirement it looks like there should be three due dates:
1) calculated due date
2) requested due date
3) existing due date

Can you please clarify with PO what each of them stands for? I assume that **requested due date** is the `dueDate` from borrowerRenew request but for the others I'm not sure. They may be either `needBefore` from transaction or `dueDate` from FOLIO Loan request

File: src/main/resources/swagger.api/schemas/d2ir/borrowerRenewDTO.json [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+    "dueDateTime": {
+      "description": "Due Date time",
+      "type": "integer"
+    }
+  },
+  "additionalProperties": false,
+  "required": [
+    "dueDateTime"
+  ]
+}
```
Comment (maryna-steshenko) [inline src/main/resources/swagger.api/schemas/d2ir/borrowerRenewDTO.json]: I suppose fields from BaseCircRequestDTO (transactionTime, patronId, patronAgencyCode, itemAgencyCode, itemId) should also be included here, but was not mentioned in the task description. Could you please clarify this with PO?

File: src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java [hunk 10]
```diff
@@ -265,6 +270,26 @@ public InnReachResponseDTO returnUncirculated(String trackingId, String centralC
...
+
+    if (calculatedDueDate >= requestedDueDate) {
+      transaction.setState(BORROWER_RENEW);
+      transactionRepository.save(transaction);
+    } else if (transactionState == ITEM_SHIPPED) {
```
Comment (maryna-steshenko) [inline src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java]: Why did you add checking `transactionState == ITEM_SHIPPED`?

Comment (s1mpleGit) [reply src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java]: If I understand correctly, documentation says that recall request from owning site to central server should be execute only if state of the transaction is `ITEM_SHIPPED`

File: src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java (hunk 8 above)
Comment (s1mpleGit) [reply src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java]: There are 2 types of dates: from request and from loan

File: src/main/resources/swagger.api/schemas/d2ir/borrowerRenewDTO.json (hunk 9 above)
Comment (s1mpleGit) [reply src/main/resources/swagger.api/schemas/d2ir/borrowerRenewDTO.json]: yep, documentation says that fields from BaseCircRequestDTO are mandatory. Added

File: src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java (hunk 10 above)
Comment (s1mpleGit) [reply src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java]: Do you mean this logic should be inside recall request?

File: src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java (hunk 3 above)
Comment (s1mpleGit) [reply src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java]: thanks for help. Fixed

File: src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java (hunk 7 above)
Comment (s1mpleGit) [reply src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java]: Thanks for help. Fixed

File: src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java (hunk 10 above)
Comment (s1mpleGit) [reply src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java]: Removed

File: src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java [hunk 11]
```diff
@@ -297,12 +306,69 @@ public InnReachResponseDTO recall(String trackingId, String centralCode, RecallD
...
+      innReachExternalService.postInnReachApi(centralCode, uri, dueDateForRecallRequest);
+      transaction.setState(RECALL);
+    } catch (Exception e) {
+      throw new CirculationException("Failed to recall request to central server: " + e.getMessage(), e);
+    }
+  }
+
+  private CheckOutResponseDTO renewLoan(InnReachTransaction transaction) {
+    var renewLoanRequestDTO = new RenewLoanRequestDTO();
+    renewLoanRequestDTO.setItemId(UUID.fromString(transaction.getHold().getItemId()));
```
Comment (dmtkachenko) [inline src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java]: should probably be
transaction.getHold().getFolioItemId()

File: src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java [hunk 12]
```diff
@@ -297,12 +306,69 @@ public InnReachResponseDTO recall(String trackingId, String centralCode, RecallD
...
+    renewLoanRequestDTO.setUserId(UUID.fromString(transaction.getHold().getPatronId()));
```
Comment (dmtkachenko) [inline src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java]: should probably be
transaction.getHold().getFolioPatronId()

File: src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java (hunk 11 above)
Comment (s1mpleGit) [reply src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java]: yep, already fixed

File: src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java (hunk 12 above)
Comment (s1mpleGit) [reply src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java]: yep, already fixed

File: src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java [hunk 13]
```diff
@@ -301,12 +311,67 @@ public InnReachResponseDTO recall(String trackingId, String centralCode, RecallD
...
   }
 
+  private void recallRequestToCentralSever(InnReachTransaction transaction, Date existingDueDate) {
+    var trackingId = transaction.getTrackingId();
+    var centralCode = transaction.getCentralServerCode();
+
+    String uri = resolveD2irCircPath(D2IR_ITEM_RECALL_OPERATION, trackingId, centralCode);
+
+    var dueDateForRecallRequest = new HashMap<>();
+    dueDateForRecallRequest.put("dueDateTime", existingDueDate);
```
Comment (oliinyko) [inline src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java]: D2IR API handles dates as UNIX epoch **seconds**. Are you sure this will work?

File: src/test/java/org/folio/innreach/controller/d2ir/InnReachCirculationControllerTest.java [hunk 14]
```diff
@@ -17,36 +31,12 @@
...
-import static org.folio.innreach.domain.entity.InnReachTransaction.TransactionState.RECALL;
-import static org.folio.innreach.domain.entity.InnReachTransaction.TransactionState.RECEIVE_UNANNOUNCED;
-import static org.folio.innreach.domain.entity.InnReachTransaction.TransactionState.RETURN_UNCIRCULATED;
-import static org.folio.innreach.domain.entity.InnReachTransaction.TransactionState.TRANSFER;
-import static org.folio.innreach.fixture.CirculationFixture.createItemShippedDTO;
-import static org.folio.innreach.fixture.CirculationFixture.createRecallDTO;
-import static org.folio.innreach.fixture.CirculationFixture.createTransactionHoldDTO;
-import static org.folio.innreach.fixture.ServicePointUserFixture.createServicePointUserDTO;
-
+import java.time.Instant;
```
Comment (oliinyko) [inline src/test/java/org/folio/innreach/controller/d2ir/InnReachCirculationControllerTest.java]: please fix import

File: src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java [hunk 15]
```diff
@@ -301,12 +311,67 @@ public InnReachResponseDTO recall(String trackingId, String centralCode, RecallD
...
+        }
+      } catch (Exception e) {
+        if (existingDueDate.before(requestedDueDate)) {
+          recallRequestToCentralSever(transaction, existingDueDate);
+        } else {
+          throw new CirculationException("Failed to renew loan: " + e.getMessage(), e);
+        }
+      }
+
+    return success();
```
Comment (oliinyko) [inline src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java]: fix code alignment

File: src/test/java/org/folio/innreach/controller/d2ir/InnReachCirculationControllerTest.java (hunk 14 above)
Comment (s1mpleGit) [reply src/test/java/org/folio/innreach/controller/d2ir/InnReachCirculationControllerTest.java]: fixed

File: src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java (hunk 15 above)
Comment (s1mpleGit) [reply src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java]: fixed

File: src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java (hunk 13 above)
Comment (s1mpleGit) [reply src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java]: added date conversion

**Review by oliinyko — APPROVED** [review]

**Review by dmtkachenko — APPROVED** [review]
